keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+
+### Use Labeling Conventions
+Rather than introducing an API object, we could require that all tools use a standard labeling scheme to identify 
+the components of an Application. This would also provide an opt-in method for tool interoperability. However, this 
+method provides no way to garbage collect Application components, express dependencies between applications, or to 
+aggregate application meta-data in a single, well-known object. Also, there is no way to ensure that tools implement 
+labeling conventions consistently or correctly. Well intentioned implementors may accidentally produce tools that 
+incorrectly implement the convention leading to degraded or broken user experiences for other tools or UIs.
+This method lacks the capacity for server side validation of the applications definition. There is not way to 
+ensure a collection of labels correctly communicates the users intentions. Lastly, this method lacks any 
+method to provide health checks.
+
+### Use Custom Resource Definitions
+Rather than introducing an API object, each tool could simply use a CRD (Custom Resource Definition) to represent 
+Applications. As CRDs are, by definition, a custom API extension mechanism, it is unlikely that this method would 
+provide for the desired level of interoperability across tools and UIs. If each tool uses their own CRD, they will a 
+priori not be interoperable, and no UI could hope to surface relevant information for all of them in a uniform manner. 
+If all tools use a common CRD, then we might as well introduce that resource as a well known Kind.
+
+### Application Type Aggregation
+An Application object must be created for each instance of application and not each type of application. It 
+might be useful to provide an object that aggregates Applications by type (e.g. all instances of WordPress that are 
+installed in the system), but this can also be accomplished by querying for all Applications and filtering or 
+aggregating based on the declared `.Spec.Type` field. In the future, if the Application Kind is widely adopted, it would 
+be possible to extend the system with another Kind that provides this aggregation via the API object model. The 
+challenges here would be multi-tenancy and namespace isolation. If a higher level object aggregates all Applications 
+of the same Type, it would have to do so across namespaces. This is not compatible with the existing namespace model 
+for Kubernetes unless we assume that the higher level object is only accessible to cluster administrators with 
+permissions to access all namespaces in the cluster.
+
+### Use the Garbage Collector Instead of an Application Controller
+We could potentially save core hours and memory by incorporating the control logic of the Application Controller 
+directly into the Garbage Collector. This design would have a poor separation of concerns and lead to poor code 
```

Comment:
Ultimately the garbage collector is a specialized graph controller.  There needs to be effective code reuse between multiple components doing generalized graph operations on large subsets of the tree, and potentially caching in the future.  I wouldn't start with the assumption it should be joined with GC, but I would start with the assumption that out of this should be common code between them.

Reply:
I'm saying the opposite here. That it would be bad design to inject more functionality into the GC controller.
I did call out under risks and mitigations that we should mirror the GC controllers implementation. The implication, for a software engineering best practices perspective,  is that we would attempt to modularize and share code between the two.